File: backend/src/main/kotlin/ru/itmo/db/coursework/api/mapping/MachineDtoMapper.kt
package ru.itmo.db.coursework.api.mapping

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import ru.itmo.db.coursework.api.dto.MachineDto
import ru.itmo.db.coursework.model.CircuitBoard
import ru.itmo.db.coursework.model.Machine
import ru.itmo.db.coursework.model.Tea

@Service
class MachineDtoMapper @Autowired constructor(
        private val machineRecipeDtoMapper: MachineRecipeDtoMapper
) {
    fun fromDto(dto: MachineDto, circuitBoardById: Map<Int, CircuitBoard>, teaById: Map<Int, Tea>) = with(dto, {
        Machine(
                id = id,
                name = name,
                sizeX = sizeX,
                sizeY = sizeY,
                price = price,
                recipes = recipes.map{
                    machineRecipeDtoMapper.fromDto(it, circuitBoardById[it.circuitBoardId]!!, teaById)
                }.toSet()
        )
    })

    fun toDto(model: Machine) = with(model, {
        MachineDto(
                id = id,
                name = name,
                sizeX = sizeX,
                sizeY = sizeY,
                price = price,
                recipes = recipes.map(machineRecipeDtoMapper::toDto).toSet()
        )
    });
}

File: frontend/src/api/mapping/MachineDtoMapper.ts
import TeaDto from "$src/api/dto/TeaDto";
import Tea from "$src/game/model/Tea";
import MachineDto from "$src/api/dto/MachineDto";
import Machine from "$src/game/model/Machine";
import MachineRecipeDtoMapper from "$src/api/mapping/MachineRecipeDtoMapper";

export default class MachineDtoMapper {
  public static fromDto(dto: MachineDto) {
    return new Machine(dto.id, dto.name, dto.size_x, dto.size_y, dto.price, dto.recipes.map(MachineRecipeDtoMapper.fromDto))
  }

  public static toDto(model: Machine) {
    return new MachineDto(model.id, model.name, model.sizeX, model.sizeY, model.price, model.recipes.map(MachineRecipeDtoMapper.toDto))
  }
}